Foundation/CPCache.j

```diff
@@ -0,0 +1,350 @@
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
+ */
+
+
+@import "CPObject.j"
+
+/*
+ * Class _CPCacheItem
+ * Represent an item of CPCache
+ * This class allow to associate a cost and a position to an object
+ *
+ * Attributes:
+ * - object: the stored object
+ * - cost: represent the cost (of memory) of the object
+ * - position: represent the insertion order to determine the oldest object
+ */
+@implementation _CPCacheItem : CPObject
+{
+    CPObject    _object      @accessors(property=object);
+    int         _cost        @accessors(property=cost);
+    int         _position    @accessors(property=position);
+}
+
++ (id)cacheItemWithObject:(CPObject)anObject cost:(int)aCost position:(int)aPosition
+{
+    var cacheItem = [[super alloc] init];
+
+    if (cacheItem)
+    {
+        cacheItem._object = anObject;
+        cacheItem._cost = aCost;
+        cacheItem._position = aPosition;
+    }
+
+    return cacheItem;
+}
+
+@end
+
+
+/*
+ * Delegate CPCacheDelegate
+ *
+ * - cache:willEvictObject: is called when a object is going to be removed
+ * When the total cost or the count exceeds the total cost limit or the count limit
+ * And also when removeObjectForKey or removeAllObjects are called
+ */
+@protocol CPCacheDelegate <CPObject>
+
+@optional
+- (void)cache:(CPCache)cache willEvictObject:(id)obj;
+
+@end
+
+
+var CPCacheDelegate_cache_WillEvictObject = 1 << 1;
+
+/*!
+    @class CPCache
+    @ingroup foundation
+    @brief A collection-like container with discardable objects
+
+    https://developer.apple.com/library/mac/documentation/Cocoa/Reference/NSCache_Class/index.html#//apple_ref/occ/instp/NSCache/delegate
+
+    A CPCache object is a collection-like container, or cache, that stores key-value pairs,
+    similar to the CPDictionary class. Developers often incorporate caches to temporarily
+    store objects with transient data that are expensive to create.
+
+    Reusing these objects can provide performance benefits, because their values do not have to be recalculated.
+    However, the objects are not critical to the application and can be discarded if memory is tight.
+    If discarded, their values will have to be recomputed again when needed.
+ */
+@implementation CPCache : CPObject
+{
+    CPDictionary            _items              @accessors(readonly);  // TOFIX: delete accessors which is only needed for test
+    int                     _currentPosition;
+    int                     _countLimit;
+    int                     _totalCostLimit;
+    id <CPCacheDelegate>    _delegate;
+
+    // Set of bit to determine which delegate has to respond
+    unsigned                _implementedDelegateMethods;
+}
+
+
+#pragma mark Initialization
+
+/*!
+    Initializes the cache with default values
+    @return the initialized cache
+*/
+- (id)init
+{
+    self = [super init];
+
+    if (self)
```

Comment:
style.

We usually do 

```
- (id)init
{
    if (self = [super init])
    {
        // init
    }

    return self;
}
```

I guess that's the only place assignation in a `if` should be tolerated in the world :)